bluecellulab/stimulus/factory.py

```diff
@@ -299,106 +300,167 @@ def ramp(
         )
 
     def ap_waveform(
-        self, threshold_current: float, threshold_percentage: float = 220.0
+        self,
+        threshold_current: Optional[float] = None,
+        threshold_percentage: Optional[float] = 220.0,
+        amplitude: Optional[float] = None,
     ) -> Stimulus:
         """Returns the APWaveform Stimulus object, a type of Step stimulus.
 
         Args:
             threshold_current: The threshold current of the Cell.
             threshold_percentage: Percentage of desired threshold_current amplification.
+            amplitude: Raw amplitude of input current.
         """
         pre_delay = 250.0
         duration = 50.0
         post_delay = 250.0
-        return Step.threshold_based(
-            self.dt,
-            pre_delay=pre_delay,
-            duration=duration,
-            post_delay=post_delay,
-            threshold_current=threshold_current,
-            threshold_percentage=threshold_percentage,
-        )
+
+        if threshold_current is not None and threshold_current != 0 and threshold_percentage is not None:
+            return Step.threshold_based(
+                self.dt,
+                pre_delay=pre_delay,
+                duration=duration,
+                post_delay=post_delay,
+                threshold_current=threshold_current,
+                threshold_percentage=threshold_percentage,
+            )
+
+        if amplitude is not None:
+            return Step.amplitude_based(
+                self.dt,
+                pre_delay=pre_delay,
+                duration=duration,
+                post_delay=post_delay,
+                amplitude=amplitude,
+            )
+
+        raise TypeError("You have to give either threshold_current or amplitude")
 
     def idrest(
         self,
-        threshold_current: float,
-        threshold_percentage: float = 200.0,
+        threshold_current: Optional[float] = None,
+        threshold_percentage: Optional[float] = 200.0,
+        amplitude: Optional[float] = None,
```

Comment:
What if the user mistakenly provides all the 3 of `threshold_current`, `threshold_percentage` and `amplitude`?
Or what if both the `amplitude` and `threshold_current` are provided but the `threshold_percentage` missing?
Can those cases also be covered with logs/exceptions/warnings?